spec/liquid_tags/unified_embed/tag_spec.rb

```diff
@@ -12,6 +28,20 @@
     expect(GistTag).to have_received(:new)
   end
 
+  it "delegates parsing to the link-matching class when there are options" do
```

Comment:
I need help with this spec. I am trying to cover the case when there's an option passed in with the embed URL. I am using the GithubTag for this spec.

When you comment out lines `39-42` below and run the spec, you get the following error, even though I believe I have stubbed the request within the `stub_github_request` method on line `4`:

```
Failures:

  1) UnifiedEmbed::Tag delegates parsing to the link-matching class when there are options
     Failure/Error: target.public_send(method, *args, &block)
     
     WebMock::NetConnectNotAllowedError:
       Real HTTP connections are disabled. Unregistered request: GET https://api.github.com/repos/forem/forem with headers {'Accept'=>'application/vnd.github.v3+json', 'Authorization'=>'Basic MDVkNjdlNjQ2MDJmZDhmOTFkNmM6NjEyYTZiMTc5NTkyOWFkZDc3NTJmNGUwOWU1Mjc2OWIxYjdmMDJjZA==', 'Content-Type'=>'application/json', 'Expect'=>'', 'User-Agent'=>'Octokit Ruby Gem 4.22.0 (http://localhost:3000)', 'X-Honeycomb-Trace'=>'1;dataset=,trace_id=11dbb11ccb6066b016eb8fa4c19c0dca,parent_id=66992e86313d9bdc,context=e30='}
     
       You can stub this request with the following snippet:
     
       stub_request(:get, "https://api.github.com/repos/forem/forem").
         with(
           headers: {
          'Accept'=>'application/vnd.github.v3+json',
          'Authorization'=>'Basic MDVkNjdlNjQ2MDJmZDhmOTFkNmM6NjEyYTZiMTc5NTkyOWFkZDc3NTJmNGUwOWU1Mjc2OWIxYjdmMDJjZA==',
          'Content-Type'=>'application/json',
          'Expect'=>'',
          'User-Agent'=>'Octokit Ruby Gem 4.22.0 (http://localhost:3000)',
          'X-Honeycomb-Trace'=>'1;dataset=,trace_id=11dbb11ccb6066b016eb8fa4c19c0dca,parent_id=66992e86313d9bdc,context=e30='
           }).
         to_return(status: 200, body: "", headers: {})
     
       registered request stubs:
     
       stub_request(:get, "https://api.github.com/repos/forem/forem").
         with(
           headers: {
          'Accept'=>'application/vnd.github.v3+json',
          'Authorization'=>'Basic MDVkNjdlNjQ2MDJmZDhmOTFkNmM6NjEyYTZiMTc5NTkyOWFkZDc3NTJmNGUwOWU1Mjc2OWIxYjdmMDJjZA==',
          'Content-Type'=>'application/json',
          'Expect'=>'',
          'User-Agent'=>'Octokit Ruby Gem 4.22.0 (http://localhost:3000)',
          'X-Honeycomb-Trace'=>'1;dataset=,trace_id=7119a1f3c49c8b7947db8f98f72f4dba,parent_id=4c72a3ec1d488f4b,context=e30='
           })
       stub_request(:head, "https://github.com/forem/forem").
         with(
           headers: {
          'Accept'=>'*/*',
          'User-Agent'=>'Ruby'
           })
       stub_request(:any, "/robohash.org/").
         with(
           headers: {
          'Accept'=>'*/*',
          'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
          'User-Agent'=>'Ruby'
           })
       stub_request(:post, "http://www.google-analytics.com/collect")
       stub_request(:any, "/dummyimage.com/")
       stub_request(:any, "/api.mailchimp.com/")
       stub_request(:get, "/twimg.com/")
       stub_request(:any, "/localhost:9090/")
       stub_request(:post, "/api.fastly.com/")
       stub_request(:any, "/emojipedia-us.s3.dualstack.us-west-1.amazonaws.com/")
       stub_request(:any, "/res.cloudinary.com/")
     
       ============================================================
     # ./app/services/github/oauth_client.rb:44:in `public_send'
     # ./app/services/github/oauth_client.rb:44:in `block in method_missing'
     # ./app/services/github/oauth_client.rb:65:in `request'
     # ./app/services/github/oauth_client.rb:43:in `method_missing'
     # ./app/liquid_tags/github_tag/github_readme_tag.rb:13:in `render'
     # ./app/liquid_tags/github_tag.rb:27:in `pre_render'
     # ./app/liquid_tags/github_tag.rb:10:in `initialize'
     # ./app/liquid_tags/unified_embed/tag.rb:47:in `new'
     # ./spec/liquid_tags/unified_embed/tag_spec.rb:51:in `block (2 levels) in <top (required)>'
     # ./spec/rails_helper.rb:146:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:146:in `block (2 levels) in <top (required)>'
     # ./spec/rails_helper.rb:101:in `block (2 levels) in <top (required)>'

Finished in 0.28659 seconds (files took 7.86 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/liquid_tags/unified_embed/tag_spec.rb:31 # UnifiedEmbed::Tag delegates parsing to the link-matching class when there are options
```

Reply:
oddly, I get a problem here because the github credentials are wrong (I have optional:optional in the authorization, while the stubbed request headers suggest a "real" key set (do you have a github api  key configured in your .env file? You probably want to invalidate it if so).

Your diff however is pointing to the honeycomb trace header's value instead (I'm not sure if we need to specify the honeycomb trace id in the stub, that feels like a lot of work).

Would using a vcr cassette (where the "ugly details" of the request headers are tucked away in a yaml file) help any here?

Reply:
suggestion: use vcr instead of stubbing the request with headers

```ruby
 it "delegates parsing to the link-matching class when there are options", vcr: true do
    link = "https://github.com/rust-lang/rust"

    allow(GithubTag).to receive(:new).and_call_original

    stub_request_head(link)
    
    VCR.use_cassette("github_client_repository_no_readme") do
      parsed_tag = Liquid::Template.parse("{% embed #{link} noreadme %}")

      expect { parsed_tag.render }.not_to raise_error
      expect(GithubTag).to have_received(:new)
    end
  end